test/adaptive_load/step_controller_test.cc

```diff
@@ -0,0 +1,194 @@
+#include "api/adaptive_load/input_variable_setter_impl.pb.h"
+#include "api/adaptive_load/step_controller_impl.pb.h"
+#include "api/client/options.pb.h"
+
+#include "adaptive_load/plugin_loader.h"
+#include "adaptive_load/step_controller_impl.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+
+namespace {
+
+nighthawk::adaptive_load::BenchmarkResult MakeBenchmarkResultOutsideThreshold() {
+  nighthawk::adaptive_load::BenchmarkResult result;
+  nighthawk::adaptive_load::MetricEvaluation* evaluation = result.add_metric_evaluations();
+  evaluation->set_threshold_score(-1.0);
+  evaluation->set_weight(10.0);
+  return result;
+}
+
+nighthawk::adaptive_load::BenchmarkResult MakeBenchmarkResultWithinThreshold() {
+  nighthawk::adaptive_load::BenchmarkResult result;
+  nighthawk::adaptive_load::MetricEvaluation* evaluation = result.add_metric_evaluations();
+  evaluation->set_threshold_score(1.0);
+  evaluation->set_weight(10.0);
+  return result;
+}
+
+nighthawk::adaptive_load::BenchmarkResult MakeBenchmarkResultWithNighthawkError() {
+  nighthawk::adaptive_load::BenchmarkResult result;
+  result.mutable_status()->set_code(1);
+  return result;
+}
+
+TEST(ExponentialSearchStepControllerConfigFactory, GeneratesEmptyConfigProto) {
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.exponential-search");
+  Envoy::ProtobufTypes::MessagePtr message = config_factory.createEmptyConfigProto();
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig expected_config;
+  EXPECT_EQ(message->DebugString(), expected_config.DebugString());
+}
+
+TEST(ExponentialSearchStepControllerConfigFactory, CreatesCorrectFactoryName) {
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  nighthawk::client::CommandLineOptions options;
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.exponential-search");
+  EXPECT_EQ(config_factory.name(), "nighthawk.exponential-search");
+}
+
+TEST(ExponentialSearchStepControllerConfigFactory, CreatesCorrectPluginType) {
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  nighthawk::client::CommandLineOptions options;
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.exponential-search");
+  StepControllerPtr plugin = config_factory.createStepController(config_any, options);
+  EXPECT_NE(dynamic_cast<ExponentialSearchStepController*>(plugin.get()), nullptr);
+}
+
+TEST(ExponentialSearchStepController, UsesInitialRps) {
+  nighthawk::adaptive_load::ExponentialSearchStepControllerConfig config;
+  config.set_initial_value(100.0);
```

Comment:
InputVariableSetter is a one-size-fits-all interface, so it always uses double. RequestsPerSecondInputVariableSetter casts the value to int and sets the integer proto field, so 100.5 will become 100.

I think the exponential search controller will actually routinely compute suggest non-integer values. It should all be ok since every InputVariableSetter is explicitly written to be able to deal with a double. Some future InputVariableSetter could want a precise fractional value